velox/functions/prestosql/window/benchmarks/PrefixSortBenchmark.cpp

```diff
@@ -0,0 +1,334 @@
+
+    std::vector<RowVectorPtr> vectors;
+    for (auto i = 0; i < kNumVectors; ++i) {
+      std::vector<VectorPtr> children;
+
+      // Generate key with a small number of unique values from a small range
+      // (0-16).
+      children.emplace_back(makeFlatVector<int32_t>(
+          kRowsPerVector, [](auto row) { return row % 17; }));
+
+      // Generate key with a small number of unique values from a large range
+      // (300 total values).
+      children.emplace_back(
+          makeFlatVector<int32_t>(kRowsPerVector, [](auto row) {
+            if (row % 3 == 0) {
+              return std::numeric_limits<int32_t>::max() - row % 100;
+            } else if (row % 3 == 1) {
+              return row % 100;
+            } else {
+              return std::numeric_limits<int32_t>::min() + row % 100;
+            }
+          }));
+
+      // Generate key with many unique values from a large range (500K total
+      // values).
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+
+      // Generate a column with increasing values to get a deterministic sort
+      // order.
+      children.emplace_back(makeFlatVector<int32_t>(
+          kRowsPerVector, [](auto row) { return row; }));
+
+      // Generate random values without nulls.
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+      children.emplace_back(fuzzer.fuzzFlat(BIGINT()));
+      children.emplace_back(fuzzer.fuzzFlat(REAL()));
+      children.emplace_back(fuzzer.fuzzFlat(DOUBLE()));
+
+      // Generate random values with nulls.
+
+      opts.nullRatio = 0.05; // 50%
+      fuzzer.setOptions(opts);
+
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+      children.emplace_back(fuzzer.fuzzFlat(BIGINT()));
+      children.emplace_back(fuzzer.fuzzFlat(REAL()));
+      children.emplace_back(fuzzer.fuzzFlat(DOUBLE()));
+
+      vectors.emplace_back(makeRowVector(inputType_->names(), children));
+    }
+    filePath_ = TempFilePath::create();
+    writeToFile(filePath_->getPath(), vectors);
+  }
+
+  ~WindowBenchmark() override {
+    HiveConnectorTestBase::TearDown();
+  }
+
+  void TestBody() override {}
+
+  // Enhance this function to include a frame clause.
+  void run(
+      const std::string& key,
+      const std::string& aggregate,
+      bool prefixSort = true) {
+    folly::BenchmarkSuspender suspender;
+
+    std::string functionSql = fmt::format(
+        "{} over (partition by {} order by k_sort)", aggregate, key);
+    core::PlanFragment plan = PlanBuilder()
+                                  .tableScan(inputType_)
+                                  .window({functionSql})
+                                  .planFragment();
+
+    vector_size_t numResultRows = 0;
+    auto task = makeTask(plan, prefixSort);
+
+    task->addSplit(
+        "0", exec::Split(makeHiveConnectorSplit(filePath_->getPath())));
+    task->noMoreSplits("0");
+
+    suspender.dismiss();
+
+    while (auto result = task->next()) {
```

Comment:
It looks like this benchmark measures combined perf of table scan + window. What's the breakdown of cpu time between scan and window? What speedup do we see for the window alone?

Reply:
Have revised the benchmark to use values instead of TableScan now and print out the timings for Window operator. The improvements with prefix-sort are modest https://gist.github.com/aditi-pandit/4942d49d91394343f57010bc72d6c831

Reply:
According to the previous benchmark of PrefixSort, in the 1k-10k scenario, even a single column has performance improvement.
see ：https://github.com/facebookincubator/velox/pull/8146
StdSort_no-payload_1_bigint_0.1k                           87.39ns    11.44M
PrefixSort                                      99.499%    87.83ns    11.39M
StdSort_no-payload_2_bigint_0.1k                          139.93ns     7.15M
PrefixSort                                      162.24%    86.25ns    11.59M
StdSort_no-payload_3_bigint_0.1k                          186.27ns     5.37M
PrefixSort                                      186.72%    99.76ns    10.02M
StdSort_no-payload_4_bigint_0.1k                          234.01ns     4.27M
PrefixSort                                      187.97%   124.49ns     8.03M
StdSort_no-payloads_1_bigint_1k                           173.31ns     5.77M
PrefixSort                                      136.72%   126.76ns     7.89M
StdSort_no-payloads_2_bigint_1k                           249.77ns     4.00M
PrefixSort                                      199.49%   125.20ns     7.99M
StdSort_no-payloads_3_bigint_1k                           314.18ns     3.18M
PrefixSort                                      219.49%   143.14ns     6.99M
StdSort_no-payloads_4_bigint_1k                           348.38ns     2.87M
PrefixSort                                      203.28%   171.38ns     5.84M
StdSort_no-payloads_1_bigint_10k                          251.90ns     3.97M
PrefixSort                                      165.99%   151.76ns     6.59M
StdSort_no-payloads_2_bigint_10k                          363.09ns     2.75M
PrefixSort                                      253.07%   143.47ns     6.97M
StdSort_no-payloads_3_bigint_10k                          483.58ns     2.07M
PrefixSort                                      293.67%   164.67ns     6.07M
StdSort_no-payloads_4_bigint_10k                          593.29ns     1.69M

And the data pattern is random， I'm not sure if I understand it correctly. In the window benchmark, do we use sorted data?

···
      // Generate a column with increasing values to get a deterministic sort
      // order.
      children.emplace_back(makeFlatVector<int32_t>(
          kRowsPerVector, [](auto row) { return row; }));
···

Reply:
Thanks for these pointers. 

The sorting keys in window are {partition by columns + order by columns}. So we have a prefix of sorting keys that are random. The last order-by-key was added for deterministic results for verification purposes. But we could remove it for the benchmark. Let me see if that improves the results.

Let me also increase the data set size and check the results.

Reply:
@skadilover : I did more runs now 
i) Unique k_sort column : https://gist.github.com/aditi-pandit/32b64b57a7167cc8d1963ef6b7edb6bd
ii) Random k_sort column : https://gist.github.com/aditi-pandit/c5e1d88ac400ce6583e2bfa6f26321c1
iii) Increase kNumRowsPerVector 10000 * 50 rows : https://gist.github.com/aditi-pandit/25d2f6224c97d9d9a32cdd97e05528fa

The run with the bigger number of rows shows most improvement, even around 30% in some cases.

In general, we can go with enabling prefixSort in Window operator

Reply:
@aditi-pandit  I think the 'NDV' of  partition by columns will affect performance：

If their NDV is large, sorting is equivalent to sorting only the partition columns, which does not reflect the advantages of binary-string-compare.

And I think in a real query, the NDV of the partition column is always relatively small.

Reply:
@skadilover : You are right... We have a smaller partition column + large sort column in this test. The end outcome of my tests seems like both smaller and larger sets for sort key values had similar results and aligned with NDVs of partition column. So we are good for now.